File: editions/2/js/matchmaker/main.js
import { inspect } from 'node:util';
import { electHost } from './electHosts.js';
import { balancedGames } from './balancingGames.js';
import { defaultPlayers } from './players.js';

const withIndex = (list) => list.map((item, i) => [i + 1, item]);

const prettyPrint = (value) =>
  console.log(inspect(value, { depth: null, colors: true, compact: 3 }));

const printGameInformation = ([index, game]) => {
  process.stdout.write(`Game ${index}: `);
  prettyPrint(game);
  process.stdout.write('With host: ');
  prettyPrint(electHost(game));
};

export const createDeathMatch = () => {
  console.log('Creating a balanced deathmatch\n');
  withIndex(balancedGames(defaultPlayers)).forEach(printGameInformation);
};

createDeathMatch();

File: editions/2/haskell/MatchMaker/Main.hs
import ElectHosts
import Players
import BalancingGames
import Tests
import Test.Hspec
import Text.Pretty.Simple 
import Control.Monad.IO.Class

createDeathMatch :: IO ()
createDeathMatch = do 
  putStrLn "Creating a balanced deathmatch\n"
  mapM_ printGameInformation indexedGames
  where
    indexedGames = withIndex $ balancedGames $ defaultPlayers

withIndex :: [a] -> [(Int, a)]
withIndex a = zip [1..length a] a

printGameInformation :: (Int, Game) -> IO ()
printGameInformation (index, game) = do
    putStr $ "Game " ++ show index ++ ": " 
    prettyPrint game
    putStr "With host: "
    prettyPrint $ electHost game


main :: IO ()
main = createDeathMatch

prettyPrint  :: (MonadIO m, Show a) => a -> m ()
prettyPrint = pPrintOpt CheckColorTty defaultOutputOptionsLightBg {outputOptionsCompact = True, outputOptionsCompactParens = True}
